flink-table/flink-table-storage/src/main/java/org/apache/flink/table/storage/filestore/lsm/compact/SortMergeReader.java

```diff
@@ -0,0 +1,182 @@
+        this.userKeyComparator = userKeyComparator;
+        this.accumulator = accumulator;
+
+        this.minHeap =
+                new PriorityQueue<>(
+                        (e1, e2) -> {
+                            int result = userKeyComparator.compare(e1.kv.key(), e2.kv.key());
+                            if (result != 0) {
+                                return result;
+                            }
+                            return Long.compare(e1.kv.sequenceNumber(), e2.kv.sequenceNumber());
+                        });
+        this.polled = new ArrayList<>();
+    }
+
+    @Nullable
+    @Override
+    public RecordIterator readBatch() throws IOException {
+        for (RecordReader reader : nextBatchReaders) {
+            while (true) {
+                RecordIterator iterator = reader.readBatch();
+                if (iterator == null) {
+                    // no more batches, permanently remove this reader
+                    reader.close();
+                    break;
+                }
+                KeyValue kv = iterator.next();
+                if (kv == null) {
+                    // empty iterator, clean up and try next batch
+                    iterator.releaseBatch();
+                } else {
+                    // found next kv
+                    minHeap.offer(new Element(kv, iterator, reader));
+                    break;
+                }
+            }
+        }
+        nextBatchReaders.clear();
+
+        return minHeap.isEmpty() ? null : new SortMergeIterator();
+    }
+
+    @Override
+    public void close() throws IOException {
+        for (RecordReader reader : nextBatchReaders) {
+            reader.close();
+        }
+        for (Element element : minHeap) {
+            element.iterator.releaseBatch();
+            element.reader.close();
+        }
+        for (Element element : polled) {
+            element.iterator.releaseBatch();
+            element.reader.close();
+        }
+    }
+
+    /** The iterator iterates on {@link SortMergeReader}. */
+    private class SortMergeIterator implements RecordIterator {
+        private boolean released = false;
+
+        @Override
+        public KeyValue next() throws IOException {
+            Preconditions.checkState(
+                    !released, "SortMergeIterator#advanceNext is called after release");
+            Preconditions.checkState(
+                    nextBatchReaders.isEmpty(),
+                    "SortMergeIterator#advanceNext is called even if the last call returns null. "
+                            + "This is a bug.");
+
+            // add previously polled elements back to priority queue
+            for (Element element : polled) {
+                KeyValue nextKv = element.iterator.next();
+                if (nextKv != null) {
+                    // still kvs left, add back to priority queue
+                    minHeap.offer(new Element(nextKv, element.iterator, element.reader));
```

Comment:
Can we provide a copy to `Element` for avoiding new object?

Reply:
This is possible. However we have to make sure that `Element`s in heap is not modified. Otherwise the priority queue structure will be broken.